File: agents/csharp/Core/Scheduler.cs
using System.Globalization;

namespace PhantomAgent.Core;

public class Scheduler
{
    public double SleepTime { get; private set; }
    public double JitterPercent { get; private set; }
    private DateTime? _killDate;

    private static readonly string[] DateFormats =
    {
        "yyyy-MM-ddTHH:mm:ss",
        "yyyy-MM-dd HH:mm:ss",
        "yyyy-MM-dd",
    };

    public Scheduler(double sleepTime = 5.0, double jitterPercent = 20.0, string? killDate = null)
    {
        SleepTime = sleepTime;
        JitterPercent = jitterPercent;
        _killDate = ParseDate(killDate);
    }

    public void Wait()
    {
        double jitter = SleepTime * (JitterPercent / 100.0);
        double actual = SleepTime + (Random.Shared.NextDouble() * 2 - 1) * jitter;
        actual = Math.Max(0.1, actual);
        Thread.Sleep((int)(actual * 1000));
    }

    public bool CheckKillDate()
    {
        if (_killDate == null) return false;
        return DateTime.UtcNow >= _killDate.Value;
    }

    public void UpdateConfig(double? sleep = null, double? jitter = null, string? killDate = null)
    {
        if (sleep.HasValue)
            SleepTime = Math.Max(0.1, sleep.Value);
        if (jitter.HasValue)
            JitterPercent = Math.Clamp(jitter.Value, 0.0, 100.0);
        if (killDate != null)
            _killDate = ParseDate(killDate);
    }

    private static DateTime? ParseDate(string? dateStr)
    {
        if (string.IsNullOrEmpty(dateStr)) return null;
        if (DateTime.TryParseExact(dateStr, DateFormats, CultureInfo.InvariantCulture,
                                    DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                                    out var dt))
            return dt;
        return null;
    }
}

File: agents/python/core/scheduler.py
import time
import random
import logging
from datetime import datetime

log = logging.getLogger("phantom.core.scheduler")


class Scheduler:
    """Beacon timing control with jitter and kill date."""

    def __init__(self, sleep_time: float = 5.0, jitter_percent: float = 20.0,
                 kill_date: str | None = None):
        self.sleep_time = sleep_time
        self.jitter_percent = jitter_percent
        self._kill_date = self._parse_date(kill_date) if kill_date else None

    def wait(self):
        """Sleep for configured interval with random jitter."""
        jitter = self.sleep_time * (self.jitter_percent / 100.0)
        actual = self.sleep_time + random.uniform(-jitter, jitter)
        actual = max(0.1, actual)
        log.debug(f"Sleeping {actual:.1f}s (base={self.sleep_time}, jitter={self.jitter_percent}%)")
        time.sleep(actual)

    def check_kill_date(self) -> bool:
        """Returns True if agent should self-destruct."""
        if not self._kill_date:
            return False
        return datetime.utcnow() >= self._kill_date

    def update_config(self, sleep: float | None = None, jitter: float | None = None,
                      kill_date: str | None = None):
        if sleep is not None:
            self.sleep_time = max(0.1, sleep)
        if jitter is not None:
            self.jitter_percent = max(0.0, min(100.0, jitter))
        if kill_date is not None:
            self._kill_date = self._parse_date(kill_date)
        log.info(f"Config updated: sleep={self.sleep_time}s, jitter={self.jitter_percent}%")

    @staticmethod
    def _parse_date(date_str: str) -> datetime | None:
        for fmt in ("%Y-%m-%dT%H:%M:%S", "%Y-%m-%d %H:%M:%S", "%Y-%m-%d"):
            try:
                return datetime.strptime(date_str, fmt)
            except ValueError:
                continue
        return None
